r/R/type.R

```diff
@@ -423,6 +423,24 @@ as_type <- function(type, name = "type") {
   type
 }
 
+canonical_type_str <- function(type_str) {
```

Comment:
Should `as_type()` also wrap this and accept a string type name, returning a DataType? (I know I wrote the function basically to handle `double()`, but looking at the function name I'd kinda expect it to take the string name too.) If so, you could simplify `nse_funcs$is()`.

Reply:
That occurred to me too, and I like the idea if the TODOs below can be solved. I'll work on that.

Reply:
This turned out too be more complicated than anticipated, for two reasons:
- `canonical_type_str()` canonicalizes data type strings to match the output of `DataType$ToString()`, but if we wanted to accept data type strings in `as_type()`, we would instead need to convert them to the names of the R data type functions. This is basically like the opposite of what `canonical_type_str()` does so would require adding another function.
- For data types that take arguments, recognizing string representations of them and converting them to R data type function calls would be quite difficult.

For these reasons, I think we should save this `as_type()` improvement for another time.

Reply:
ARROW-12908